### ISSUE TITLE
Jakarta EE 11:  Support Jakarta Authorization 3.0

### ISSUE THREAD

**Issue by jhanders34** [OpenLiberty/open-liberty#25420]
## Description

Implement the Jakarta Authorization 3.0 specification

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Documents

When available, add links to required feature documents. Use "N/A" to mark particular documents which are not required by the feature.

- Externally raised requests for enhancements: 
  - [Aha](https://cloud-platform.ideas.aha.io/): Link to the Aha idea (also add a link to this issue in the Aha idea)
    - Feature owner adds label `Aha idea`
  - [Open Liberty Feature Request](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=enhancement&template=feature_request.md&title=): Link to the OL GH issue (also add a link to this issue in the Feature Request GH issue)
    - Feature owner adds label `Requested feature`
- UFO: Link to [Upcoming Feature Overview](https://ibm.box.com/s/35csohjqwmitdropfoknoeev7tosyfwe) document
- FTS: #29464
- Beta Blog: https://github.com/OpenLiberty/open-liberty/issues/34469
- GA Blog: Link to GA Blog Post GH Issue

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## Process Overview

- [Prioritization](#prioritization)
- [Design](#design)
- [Implementation](#implementation)
- [Legal and Translation](#legal-and-translation)
- [Beta](#beta)
- [GA](#ga)
  - [Focal Point Approvals](#focal-point-approvals-complete-by-feature-complete-date)
- [Other Deliverables](#other-deliverables)

### General Instructions

The process steps occur roughly in the order as presented. Process steps occasionally overlap.

Each process step has a number of tasks which must be completed or must be marked as not applicable ("N/A").

Unless otherwise indicated, the tasks are the responsibility of the Feature Owner or a Delegate of the Feature Owner.

If you need assistance, reach out to the [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect).

**Important: Labels are used to trigger particular steps and must be added as indicated.** 

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Prioritization** (Complete Before Development Starts)

The ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) and area leads are responsible for prioritizing the features and determining which features are being actively worked on.

### **Prioritization**
- [x] Feature added to the "New" column of the [Open Liberty project board](https://github.com/orgs/OpenLiberty/projects/2)
  - Epics can be added to the board in one of two ways:
    - From this issue, use the "Projects" section to select the appropriate project board.
    - From the appropriate project board click "Add card" and select your Feature Epic issue
- [x] Priority assigned
  - Attend the Liberty Backlog Prioritization meeting

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Design** (Complete Before Development Starts)

Design preliminaries determine whether a formal design, which will be provided by an [Upcoming Feature Overview (UFO)](https://ibm.box.com/v/UFO-Template) document, must be created and reviewed.  A formal design is required if the feature requires any of the following: UI, Serviceability, SVT, Performance testing, or non-trivial documentation/ID.  Furthermore, each identified item places a blocking requirement on another team so it must be identified early in the process.  The feature owner may check-off the item if they know it doesn't apply, but otherwise they should work with the focal point to determine what work, if any, will be necessary and make them aware of it.

### **Design Preliminaries**
- [x] UI requirements identified, or N/A. (Feature owner and [UI focal point](https://github.com/orgs/OpenLiberty/teams/ui-approvers))
- [x] Accessibility requirements identified, or N/A. (Feature owner and [Accessibility focal point](https://github.com/orgs/OpenLiberty/teams/accessibility-approvers))
- [x] ID requirements identified, or N/A. (Feature owner and [ID focal point](https://github.com/orgs/OpenLiberty/teams/id-approvers))
   - Refer to [Documenting Open Liberty](https://github.com/OpenLiberty/open-liberty/wiki/Documenting-Open-Liberty).
   - Feature Owner adds label `ID Required`, if non-trivial documentation needs to be created by the ID team.
   - ID adds label `ID Required - Trivial`, if no design will be performed and only trivial ID updates are needed.
- [x] Serviceability requirements identified, or N/A. (Feature owner and [Serviceability focal point](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
- [x] SVT requirements identified, or N/A. (Feature owner and [SVT focal point](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
- [x] Performance testing requirements identified, or N/A. (Feature owner and [Performance focal point](https://github.com/orgs/OpenLiberty/teams/performance-approvers))

### **Design**
- [x] POC Design / UFO review scheduled.
  - Follow the instructions in POC-Forum repo
- [x] POC Design / UFO review completed.
- [x] POC / UFO Review follow-ons completed.
- [x] POC Design / UFO approval requested.
  - Feature owner adds label `Design Approval Request`
- [x] Design / UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) adds label `Design Approved`
  - Add the public link to the UFO in Box to the [Documents](#documents) section.
  - The UFO must always accurately reflect the final implementation of the feature. Any changes must be first approved. Afterwards, update the UFO by creating a copy of the original approved slide(s) at the end of the deck and prepend "OLD" to the title(s). A single updated copy of the slide(s) should take the original's place, and have its title(s) prepended with "UPDATED".

### **No Design**
- [ ] No Design requested.
  - Feature owner adds label `No Design Approval Request`
- [x] No Design / No UFO approved. ([OpenLiberty/chief-architect](https://github.com/orgs/OpenLiberty/teams/chief-architect)) or N/A
  - Approver adds label `No Design Approved`
- [x] Feature / Capability stabilization or discontinuation or N/A
  - Feature owner adds label `Product Management Approval Request` and notifies [OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management)
  - Approver adds label `Product Management Approved` ([OpenLiberty/product-management](https://github.com/orgs/OpenLiberty/teams/product-management))
  - Note: For stabilized, superseded, and discontinued feature/capability, skip the [Beta](#beta) section of the template (you may delete it).  Otherwise, proceed as normal.

### **FAT Documentation**
- [x] "Feature Test Summary" child task created
  - Use the [Feature Test Summary Template](https://github.com/OpenLiberty/open-liberty/issues/new?assignees=&labels=Feature+Test+Summary&template=feature_test_summary.md&title=)
  - Add FTS issue link to the [Documents](#documents) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Implementation**

A feature must be [prioritized](https://github.com/orgs/OpenLiberty/projects/2) before any implementation work may begin to be delivered (inaccessible/no-ship).  However, a design focused approach should still be applied to features, and developers should think about the feature design prior to writing and delivering any code.  
Besides being prioritized, a feature must also be socialized (or No Design Approved) before any beta code may be delivered.  All new Liberty content must be inaccessible in our GA releases until it is [Feature Complete](#feature-complete) by either marking it `kind=noship` or [beta fencing](#beta-code) it.  
Code may not GA until this feature has obtained the `Design Approved` or `No Design Approved` label, along with all other tasks outlined in the [GA](#ga) section.

### **Feature Development Begins**
- [x] Add the `In Progress` label

## **Legal and Translation**

In order to avoid last minute blockers and significant disruptions to the feature, the **legal items need to be done as early in the feature process as possible**, either in design or as early into the development as possible.  Similarly, translation is to be done concurrently with development.  Both **MUST** be completed before Beta or GA is requested.

### **Legal** (Complete before Feature Complete Date)
- [x] Changed or new open source libraries are cleared and approved, or N/A. (Legal Release Services/Cass Tucker/Release PM).

### **Innovation** (Complete 1 week before Feature Complete Date)
- [x] Consider whether any aspects of the feature may be patentable. If any identified, disclosures have been submitted.

### **Translation** (Complete by Feature Complete Date)
- [x] PII (Program Integrated Information) updates are merged (i.e. all English strings due for translation have been delivered), or N/A.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Beta**

In order to facilitate early feedback from users, all new features and functionality should first be released as part of a beta release.

### **Beta Code**
- [x] [Beta fence](https://github.com/OpenLiberty/open-liberty/wiki/Beta-Fencing) the functionality
  - E.g. `kind=beta`, `ibm:beta`, `ProductInfo.getBetaEdition()`
- [x] Beta development complete and feature ready for inclusion in a beta release
  - Add label `target:beta` and the appropriate `target:YY00X-beta` (where YY00X is the targeted beta version).
- [x] Feature delivered into beta
  - ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager)) adds label `release:YY00X-beta` (where YY00X is the first beta version that included the functionality).

### **Beta Blog** (Complete by beta eGA)
- [x] Beta blog issue created and populated using the [Open Liberty BETA blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the beta blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [beta release blog post](https://openliberty.io/blog/?search=beta&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **GA**

A feature is ready to GA after it is Feature Complete and has obtained all necessary Focal Point Approvals. 

### **Feature Complete**
- [ ] Feature implementation and tests completed.
  - [ ] All PRs are merged.
  - [ ] All related/child issues are closed.
  - [ ] All stop ship issues are completed.
- [x] Legal: all necessary approvals granted.
- [x] Translation: Feature may only proceed to GA if it has either `Translation - Complete` or `Translation - Missing` label
  - If all translation has been delivered to `release` branch, feature owner adds label `Translation - Complete`.
  - If missing translation does not cause a break in functionality, nor a security or production outage risk, feature owner adds label `Translation - Missing`.
    - Once all missing translations are delivered, the `Translation - Missing` label is replaced with `Translation - Complete`.
  - If missing translation could cause a break in functionality or a security or production outage risk, feature owner adds the `Translation - Blocked` label.
    - Featues with `Translation - Blocked` may **NOT** proceed to GA until the label has been replaced with either `Translation - Missing` or `Translation - Complete`.
  - For further guidance, contact [Globalization focal point](https://github.com/orgs/OpenLiberty/teams/globalization-approvers) or the [Release Architect](https://github.com/orgs/OpenLiberty/teams/release-architect).
- [ ] GA development complete and feature ready for inclusion in a GA release
  - Add label `target:ga` and the appropriate `target:YY00X` (where YY00X is the targeted GA version).
  - Inclusion in a release requires the completion of all Focal Point Approvals.

### **Focal Point Approvals** (Complete by Feature Complete Date)

These occur only after GA of this feature is requested (by adding a `target:ga` label).  GA of this feature may not occur until all approvals are obtained.

### **All Features**
- [ ] **APIs/Externals** - Externals have been reviewed or N/A.  ([OpenLiberty/externals-approvers](https://github.com/orgs/OpenLiberty/teams/externals-approvers))
  - Approver adds label `focalApproved:externals` 
- [ ] **Demo** - Demo is scheduled for an upcoming EOI or N/A. ([OpenLiberty/demo-approvers](https://github.com/orgs/OpenLiberty/teams/demo-approvers))
  - Add comment `@OpenLiberty/demo-approvers Demo scheduled for EOI [Iteration Number]` to this issue.
  - Approver adds label `focalApproved:demo`.
- [ ] **FAT** - All Tests complete and running successfully in SOE or N/A. ([OpenLiberty/fat-approvers](https://github.com/orgs/OpenLiberty/teams/fat-approvers))
  - Approver adds label `focalApproved:fat`.

### **Design Approved Features**
- [ ] **ID** - Documentation is complete or N/A. ([OpenLiberty/id-approvers](https://github.com/orgs/OpenLiberty/teams/id-approvers))
  - Approver adds label `focalApproved:id`.
  - > **_NOTE:_**  If only trivial documentation changes are required, you may reach out to the ID Feature Focal to request a `ID Required - Trivial` label.  Unlike features with regular ID requirement, those with `ID Required - Trivial` label do not have a hard requirement for a Design/UFO.
- [x] **InstantOn** - InstantOn capable or N/A. ([OpenLiberty/instantOn-approvers](https://github.com/orgs/OpenLiberty/teams/instantOn-approvers))
  - Approver adds label `focalApproved:instantOn`.
- [x] **Performance** - Performance testing is complete or N/A. ([OpenLiberty/performance-approvers](https://github.com/orgs/OpenLiberty/teams/performance-approvers))
  - Approver adds label `focalApproved:performance`.
- [ ] **Serviceability** - Serviceability has been addressed or N/A. ([OpenLiberty/serviceability-approvers](https://github.com/orgs/OpenLiberty/teams/serviceability-approvers))
  - Approver adds label `focalApproved:sve`.
- [ ] **STE** - Skills Transfer Education chart deck is complete or N/A. ([OpenLiberty/ste-approvers](https://github.com/orgs/OpenLiberty/teams/ste-approvers))
  - Approver adds label `focalApproved:ste`.
- [ ] **SVT** - System Verification Test is complete or N/A. ([OpenLiberty/svt-approvers](https://github.com/orgs/OpenLiberty/teams/svt-approvers))
  - Approver adds label `focalApproved:svt`.

### **Remove Beta Fencing** (Complete by Feature Complete Date)
- [ ] Beta guards are removed, or N/A
  - Only after all necessary Focal Point Approvals have been granted.

### **GA Blog** (Complete by Friday after GM)
- [ ] GA Blog issue created and populated using the [Open Liberty GA release blog post](https://github.com/OpenLiberty/open-liberty/issues/new/choose) template.
  - Add a link to the GA Blog issue in the [Documents](#documents) section.
  - Note: This is for inclusion into the overall [release blog post](https://openliberty.io/blog/?search=release&key=tag).  If, in addition, you'd also like to create a dedicated blog post about your feature, then follow the "Standalone Feature Blog Post" instructions under the [Other Deliverables](#other-deliverables) section.

### **Post GM** (Complete before GA)
- [ ] After confirming this feature has been included in the GM driver, feature owner closes this issue.

### **Post GA**
- [ ] Remove the `target:ga` and `target:YY00X` labels, and add the appropriate `release:YY00X`. ([OpenLiberty/release-manager](https://github.com/orgs/OpenLiberty/teams/release-manager))

- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
## **Other Deliverables**

- [x] **Standalone Feature Blog Post** - A blog post specifically about your feature or N/A.  (Feature owner and [OpenLiberty/release-architect](https://github.com/orgs/OpenLiberty/teams/release-architect))
  - This should be strongly considered for larger or more prominent features.
  - Follow [instructions](https://github.com/OpenLiberty/blogs/tree/draft#writing-and-publishing-blog-posts-on-the-openlibertyio-blog) in the blogs repo.
- [ ] **OL Guides** - OL Guides assessment is complete or N/A. ([OpenLiberty/guide-assessment](https://github.com/orgs/OpenLiberty/teams/guide-assessment/members))
- [ ] **Dev Experience** - Developer Experience & Tools work is complete or N/A. ([OpenLiberty/dev-experience-assessment](https://github.com/orgs/OpenLiberty/teams/dev-experience-assessment/members))
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -

**Comment by njr-11** [comment]
Review comments from the Upcoming Feature Overview presentation:

- [x] Slide 15 - Check if the Permission class is being deprecated in Java SE
  I confirmed that the Permission class is not deprecated.
- [ ] Slide 38 - Note that updates are needed to Transformation Advisor
- [x] Concerns about Authorization / JACC features before 3.0 working with Java 24 due to Policy class changes
  Confirmed that JACC / Authorization features before 3.0 do not work with Java 24.  Created PR #30322 to address the problem
- [ ] Concerns about application being able to set the PolicyFactory and override the one that was set by Liberty from authorization provider's user feature

**Comment by njr-11** [comment]
Review comments from part 3 of the Upcoming Feature Overview:

- [x] Slide 6 - make sure bullet point 4 is updated to latest version of slides
- [x] Slide 24 - note when the configuration shown is unchanged
Move to Technical Background
- [x] Slide 40 - communicate need for migration to stack products?  And consider if there is anywhere the change can be documented to users.

**Comment by ayoho** [comment]
## Meeting Summary — February 27, 2026
- Jakarta Authorization 3.0 (Part 3) — Open Liberty Epic 25420

---

### Discussion & Feedback

**Slide accuracy / stale content (Nathan Rauh noted at the end):**
- **Slide 6, bullet 4** — needs to be updated to the latest version (Jared had accidentally opened an older version of the deck after rebooting).
- **Slide 24** (and the slide before it) — should note when configuration shown is *unchanged* from previous versions, to avoid confusion. Emily Jiang was confused by the security role mapping slides, thinking they were new, when they were actually existing behavior. Jared agreed to label unchanged slides more clearly.
- **Slide 40 (migration)** — needs to communicate the need for migration to stack products and check on documentation.

**Java `Policy` deprecation and future compatibility (Thomas Watson):**
- Tom asked whether the old `java.security.Policy` was gone or still present. Jared confirmed it is still there but deprecated for removal, and that Jakarta EE 11 does not use it at all since the Security Manager is not used in EE 11.
- Tom raised a concern: once `java.security.Policy` is truly removed from the JVM, parts of the existing implementation will become non-functional, since you cannot define classes in the `java.*` namespace. Jared acknowledged this was not part of the current design and said they'd have to address it later. Tom's conclusion was that those parts will have to fail gracefully and users will need to move to newer EE versions to use later Java versions.

**XML vs. injection-based configuration (Emily Jiang):**
- Emily asked whether there was an injection-based proposal for configuring authorization modules, noting the spec is trying to move away from XML. Jared confirmed there is no injection proposal; the new mechanism is XML-based (`web.xml`) and also supports programmatic `set` methods.
- Emily suggested this could potentially be a candidate for Jakarta Config or MicroProfile Config in the future, but acknowledged it was out of scope for this feature.
- Emily also suggested that slides showing existing (unchanged) configuration should be labeled as such to avoid confusion.

**Migration tooling gap for users with custom authorization user features (Emily Jiang, Thomas Watson, Chuka Obinabo):**
- Emily raised a concern that users who have existing Liberty user features implementing the old `ProviderService` interface will hit a `ClassNotFoundException` when they upgrade to Authorization 3.0, since that interface is being removed.
- Jared confirmed there is currently nothing in the migration toolkits that will detect this and warn users. The migration slide covers it at a documentation level, but there is no automated tooling action.
- Tom pointed out that users would have to explicitly update their user feature's `tolerates` list to include the new version before they'd even encounter the problem — so it's a deliberate developer action, not a silent breakage.
- Emily's concern was about how to proactively warn users *before* they hit the error. The group agreed the best avenue is clear documentation in the "what's different" section for EE 11, and that the errors Liberty emits when the class can't be found will include a message (though not a full user action for the `ClassNotFoundException` case, since it originates in user code).
- Tom noted this is a niche scenario (few implementers of custom authorization modules) and was skeptical of investing significant development resources to improve the tooling experience for it.
- The group agreed to track this offline.

**Comment by jhanders34** [comment]
**Serviceability Approval Comment** - Please answer the following questions for Serviceability approval:

1.  Have the most likely customer‑facing problem scenarios identified in the Serviceability section of the UFO been addressed in the implementation?

- Yes.  Slide 38 of [the UFO](https://ibm.ent.box.com/file/1641423657268?s=35csohjqwmitdropfoknoeev7tosyfwe) identified the likely scenarios.  

  - Messages were added to inform the user about the authorization modules failing to be instantiated and when they are added and removed.  `io.openliberty.security.authorization.internal.tests.FactoryMessagesTests` in the `io.openliberty.security.authorization.internal.3.0_fat` project validates that FFDC and error messages come out when failing to instantiate and the add and remove messages come out when it works.
  - For the ProviderService API no longer being available, the `io.openliberty.security.authorization.internal.tests.ProviderServiceRemovalTest` test was added to validate that there is a message that comes out stating that the package that is imported in the user feature's bundle is not found as part of normal OSGi package resolution processing:  `Unresolved requirement: Import-Package: com.ibm.wsspi.security.authorization.jacc`.

2.  Test and Demo -- As part of the serviceability process we're asking feature teams to test and analyze common problem paths for serviceability and demo those problem paths to someone not involved in the development of the feature (eg. IBM Support, test team, or another development team).  
a) What problem paths were tested and demonstrated? 
b) Who did you demo to?
c) Do the people you demo'd to agree that the serviceability of the demonstrated problem scenarios is sufficient to avoid Support Cases for any problems customers are likely to encounter, or that IBM Support should be able to quickly address those problems without need to engage SMEs?
- Started server with user feature that used ProviderService that was updated to tolerate appAuthorization-3.0 and showed the message that is output indicating that the package is no longer resolvable.
- Started an application with an error starting the authorization modules and showed the FFDC and error messages that come out when that error occurs.
- Also started an application where the authorization modules do not get errors and showed the messages that are output  when the application is started and stopped.

- Demoed to Tracy Burroughs and he agreed that serviceability is sufficient for the demonstrated problem scenarios, but wanted to make sure that the ProviderService scenario OSGi error message was well documented in the Liberty documentation to help customers find the documentation if they google the specific package name not being resolved by OSGi.

3.  Which IBM Support / SME queues will handle Support Cases for this feature?  Ensure they are present in the contact reference file and in the Queue Contact Summary, and that the respective IBM Support/SME teams know they are supporting it.  Ask Prashanth Gunapalasingam if you need links or more info.

- `WAS L3: Core Security`

4. Does this feature add any new metrics or emit any new JSON events?  If yes, have you updated the JMX metrics reference list / Metrics reference list / JSON log events reference list in the Open Liberty docs?

- No

**Comment by jhanders34** [comment]
@OpenLiberty/ste-approvers the STE is uploaded for your review and approval

**Comment by chirp1** [comment]
Ram completed the following doc and Jared approved: https://github.com/OpenLiberty/docs/issues/8294

Approving this feature
